cpp/tests/utilities/identify_stream_usage.cpp

```diff
@@ -66,6 +67,34 @@ rmm::cuda_stream_view const get_default_stream()
 }  // namespace test
 #endif
 
+#ifdef STREAM_MODE_TESTING
+namespace detail {
+
+/**
+ * @brief Implementation of `cuda_stream_pool` that always returns the
+ * `cudf::test::get_default_stream()`
+ */
+class test_cuda_stream_pool : public cuda_stream_pool {
+ public:
+  rmm::cuda_stream_view get_stream() override { return cudf::test::get_default_stream(); }
+  rmm::cuda_stream_view get_stream(stream_id_type stream_id) override
+  {
+    return cudf::test::get_default_stream();
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(std::size_t count) override
+  {
+    return std::vector<rmm::cuda_stream_view>(count, cudf::test::get_default_stream());
+  }
+
+  std::size_t get_stream_pool_size() const override { return 1UL; }
+};
+
+cuda_stream_pool* create_global_cuda_stream_pool() { return new test_cuda_stream_pool(); }
```

Comment:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**🚩 issue:** ‏ When is this function called rather than the one implemented in `stream_pool.cpp`? Is the other one undefined if `STREAM_MODE_TESTING` is defined? Can you clarify the documentation? It would actually be nice to have them in the same file if this is indeed how it works. Otherwise I worry about hard-to-diagnose multiple definition linker errors. It's a little confusing that the class `test_cuda_stream_pool` is defined in a file called `identify_stream_usage.cpp`.

Reply:
Everything in this file needs to be in this file. It is compiled and linked into a separate library that is then preloaded (i.e. via `LD_PRELOAD`) in certain contexts, which allows functions defined here to override the ones defined in libcudf proper. There's more information at the bottom of the testing dev guide page: https://docs.rapids.ai/api/libcudf/stable/md_developer_guide_testing.

On that note, It would be nice to update the dev guide to mention the existence of the stream pool as well, maybe just a single sentence noting that pools are also overridden by this testing stream pool.

Reply:
OK, maybe a file-level comment explaining what this is. If this is a separate library, perhaps putting it in `cpp/tests/utilities/` is not the best?

Reply:
Yeah we could certainly work on improving that. I can look into this for a follow-up PR.